Phantom-node based XFEM

bwspenc opened the pull request:
This is a big, and probably controversial set of commits. This implements XFEM based on the phantom node method. It actually has a relatively minor impact on the rest of MOOSE.  Most of the heavy lifting is done by a mesh cutting algorithm that we named ElementFragmentAlgorithm, and which has no dependencies on MOOSE.

I'm leaving this as a big chain of commits to preserve authorship. If it's better to squash them, let me know.

bwspenc committed:
Squash all XFEM commits that touch MOOSE

bwspenc committed:
Squash all XFEM commits that touch elk

bwspenc committed:
Temporary HACK: Skip copying of stateful data to avoid segfault

bwspenc committed:
Switch to using enum to define quantity to extract with XFEMCutPlaneAux

bwspenc committed:
Work in progress to make auxKernel to identify crack tip elements

bwspenc committed:
More work on flagging crack tip elements

bwspenc committed:
Fix problems with overlays_elem() that broke check for crack tip elements

bwspenc committed:
Clear CrackTipElements in the right place

bwspenc committed:
Mark cuts by state

Modify XFEM class to include a vector of elements marked for cutting
based on state, and cut those elements.  Add an ElementUserObject that
determines whether elements should be cut based on state adds elements
to be cut to the XFEM object's vector of elements to be cut.  Still need
to modify the UserObject to mark based on something meaningful, but
the framework is in place now.

bwspenc committed:
More work on XFEMMarkerUserObject

bwspenc committed:
Add XFEMMaterialTensorMarkerUserObject.  It still needs work

bwspenc committed:
Make XFEMMaterialTensorMarkerUserObject use values from the material tensor

bwspenc committed:
Copy boundary info from parent to child elements

bwspenc committed:
Add loop to repeat solution for step until XFEM cracks are done growing

bwspenc committed:
Fix bug in adding new nodes/elements to boundaries

bwspenc committed:
Add ability to initialize new cracks on a specified boundary

bwspenc committed:
Fix a bug with crack direction if the average isn’t used

bwspenc committed:
Fix bug that was messing up crack orientations in physics-based growth

bwspenc committed:
Copy material data from parent to child elements

bwspenc committed:
Add libMesh namespace

bwspenc committed:
Add random_range parameter, change how average is calculated for marker userobject

bwspenc committed:
Replace outdated to FEProblem::updateMaterials() with advanceState()

bwspenc committed:
Update syntax for MooseEnum

bwspenc committed:
Add tests of efa in MOOSE cppuint -- they just run, don't check anything at this point

bwspenc committed:
Add inverse connectivity map to EFA

bwspenc committed:
Initial creation of fragment class

bwspenc committed:
Add copy constructor for fragment that can convert between local and global nodes

bwspenc committed:
Store the fragments themselves in the host code that uses the mesh cutter

bwspenc committed:
Add initial definition of edge class

bwspenc committed:
Use edge objects in elements

bwspenc committed:
Use edges in fragments

zzyfisherman committed:
Add 2D branching function based on geometric cuts

zzyfisherman committed:
add state_based branching

zzyfisherman committed:
add visualization function for 2D branching

zzyfisherman committed:
Correct the bug of not merging two permanent child nodes

zzyfisherman committed:
CutPlaneAction added. The auxkernel of cut planes improved

bwspenc committed:
Fix compile warning

bwspenc committed:
Fix compile error in Transient.C

bwspenc committed:
Misc cleanup

Fix double deletion of node_t objects
Make a couple of parameters optional

zzyfisherman committed:
Move XFEM code to a separate folder; replace node_t with EFAnode

zzyfisherman committed:
Rename element_t and fragment_t

zzyfisherman committed:
Move as much stuff as I can into EFAelement class to make EFA more OO

zzyfisherman committed:
Rename CutElemMesh and its private members

zzyfisherman committed:
add EFAedge::isPartialOverlap

zzyfisherman committed:
Save the uncut element ahead of a crack tip; Change the method to identify crack tip elements

zzyfisherman committed:
Improve the method combine_tip_edges() to prepare for the case when one edge has two cuts

zzyfisherman committed:
Correct the possible bug in storing crack tip elements in XFEM.C

zzyfisherman committed:
Correct the bug of introducing extra permanent nodes; Clean up unit test

zzyfisherman committed:
Generalize the code to triangle mesh

zzyfisherman committed:
Allow two cuts in element edges

zzyfisherman committed:
Combine restoreFragment and restoreEdgeIntersection to clean up

zzyfisherman committed:
Create an independent class XFEMCutElem

zzyfisherman committed:
Remove childElem from the argument list of mergeNodes()

zzyfisherman committed:
2D derived fragments and elements done

zzyfisherman committed:
add class EFAface

zzyfisherman committed:
added EFAfragment3D and EFAelement3D

zzyfisherman committed:
improve EFAedge::copy_intersection()

zzyfisherman committed:
correct a bug in combine_tip_edges(), not affect the result though

zzyfisherman committed:
remove the code of duplicating embedded nodes

zzyfisherman committed:
Modify EFA.C so we pass the unit tests

zzyfisherman committed:
A bug in shouldDuplicateCrackTipSplitElem resolved; add _general_neighbors in EFAelement

zzyfisherman committed:
add derived class XFEM_geometric_cut_2d and XFEM_square_cut

bwspenc committed:
in CutPlaneAction timestep is changed to timestep_end

bwspenc committed:
Fix compile errors with latest moose/libmesh, rename cut planes

zzyfisherman committed:
EFAedge::equivalent only compares end nodes now; Mics clean-up

zzyfisherman committed:
correct a bug in XFEM_square_cut.C

zzyfisherman committed:
finish 3D single stationary crack with bugs though

zzyfisherman committed:
Move normalize() and cross_product() into EFAfuncs.h; improve normal calculation in XFEM_square_cut.C; FaceNoce.C -> FaceNode.C

zzyfisherman committed:
add MF & direct quadrature rule, remove xfem_volfrac in solid mechanics & heat kernels,
modify ElementIntegralPostprocessor to remove contributions outside real domains

zzyfisherman committed:
Remove cross_product and dot_product in EFAfuncs.h and use standard MOOSE method instead

bwspenc committed:
Fix segfault and compiler warnings

jiangwen84 committed:
Compute j integrals with XFEM

bwspenc committed:
HACK: add dummy material to get solid_mechanics state working

Commit:
add remove_invalid_embedded in EFAfragment2D.C

Commit:
add a method to handle 3 cuts, not robust though

zzyfisherman committed:
add get_frag_edges and get_frag_faces

bwspenc committed:
Initialize solution on new nodes

bwspenc committed:
Remove traces of XFEM from HeatConduction.C

bwspenc committed:
Remove some of the XFEM-related output

bwspenc committed:
Add plastic plate xfem problem

zzyfisherman committed:
Fix the bug of isolated nodes

zzyfisherman committed:
add bool arg displaced_mesh for XFEM::get_frag_faces()

bwspenc committed:
Fix stateful material update bug

bwspenc committed:
Fix segfault with material copy in parallel

bwspenc committed:
Better fix for parallel material state issue

bwspenc committed:
Update plate xfem and contact problem

bwspenc committed:
Fix parallel issue with copying nodal data off-processor.  It works in parallel now

bwspenc committed:
Fix build errors (mostly related to new constructor)

jiangwen84 committed:
Fix parallel problem with state-marked elements

jiangwen84 committed:
added support to calucate KI for user defined points. fixed circluar cut bug. added elliptic cut

jiangwen84 committed:
moment fitting now uses libmesh's quadrature information

jiangwen84 committed:
clean up: put some private functions in 3D cut into XFEMMiscFuncs.h

jiangwen84 committed:
put xfem's staff into single input block(action)

jiangwen84 committed:
move the xfem's weights from Kernal to Assembly

jiangwen84 committed:
support multiple data set for 3D cut

jiangwen84 committed:
cleanup some comments

zzyfisherman committed:
fix the weights and jacobian for triangles, fix a bug in solve_mf()

bwspenc committed:
Cleanup to get it to compile after rebase -- mostly remove deprecated headers

jiangwen84 committed:
modifed the unit test of cut algorihm

jiangwen84 committed:
added xfem test files

jiangwen84 committed:
pass all tests

jiangwen84 committed:
add a missing mesh file

bwspenc committed:
Fix compile errors with latest rebase, resolve conflict issues with Transient.C

bwspenc committed:
Clean up and add first batch of tests

bwspenc committed:
Clean up and add remainder of xfem tests

bwspenc committed:
Fix compile errors with latest rebase.

jiangwen84 committed:
change qrule type (string to enum), check and clean xfem weights calculation.

bwspenc committed:
Major cleanup of XFEM in preparation for merge to devel branch ref #6320

Name changes
File and directory renaming
Changing interface with MOOSE for computing weights
Bringing code in compliance with MOOSE conventions
Fix includes and forward declarations
Improvements to regression and unit tests
Lots of misc. fixes

dschwen commented on framework/include/postprocessors/ElementIntegralPostprocessor.h:
```diff
@@ -46,6 +46,6 @@ class ElementIntegralPostprocessor : public ElementPostprocessor
 
   Real _integral_value;
-};
+  };
```
?

dschwen commented on framework/include/xfem/EFAFuncs.h:
```diff
@@ -0,0 +1,82 @@
+#include <vector>
+
+template <typename T>
+bool deleteFromMap(std::map<unsigned int, T*> &theMap, T* elemToDelete)
```
I'd suggest wrapping all those in a namespace

dschwen commented on framework/include/xfem/EFANode.h:
```diff
@@ -0,0 +1,50 @@
+
+#include <string>
+
+enum N_CATEGORY
```
This enum should probably not be global. Can you make it a public member of one of the classes (or the namespace suggested above)?

dschwen commented on framework/include/xfem/EFAVolumeNode.h:
```diff
@@ -0,0 +1,43 @@
+  double _eta;
+  double _zeta;
+
+  public:
```
indentation

dschwen commented on framework/include/xfem/ElementFragmentAlgorithm.h:
```diff
@@ -0,0 +1,92 @@
+
+  /**
+   * Constructor
+   **/
```
This type of comment does not add much value. Comments on all the undocumented members below would be more useful.

friedmud commented:
Can all of this be done as a `module`?  I hate to add this much compiling overhead to everyone by adding all of these new objects to MOOSE itself.

bwspenc commented:
@friedmud Yes, I'd be totally in favor of doing this as a module. We'd need to figure out the dependency issues. This code really touches the framework in 3 places:
1) The XFEM object is owned by FEProblem. That's similar to adaptivity, because the XFEM object takes care of modifying the mesh just like adaptivity.
2) There are some modifications to Transient to have XFEM modify the mesh, and to repeat the solution if the mesh is modified.
3) There is an additional method in Assembly that calls XFEM to modify the integration weights on cut elements.

It would be easy enough to handle 1 and 2 in a module -- we could make our own classes that derive from FEProblem and Transient. The one I'm not sure how to handle is item 3. We'd need some kind of interface to allow our code in a module to modify the integration weights.

I'm going to be out today, but we could discuss this tomorrow.

@dschwen Those are good suggestions, and I'll address them.  I know the code is still rough around the edges. I'd like to work out the big architectural issues first and then clean up the minor stuff.

friedmud commented:
I'll have to give some thought to 3... for now... can you take a whack at moving the rest of the stuff to modules so we can get an idea of what we're dealing with?

bwspenc commented:
Sure, I'll take a stab at moving it over to a module. Maybe the easiest way to do this is to have a really stripped down base class for that XFEM class live in the framework.

friedmud commented:
We may need grow some new abstractions / interfaces to help here too.  We just need to see what's necessary.

bwspenc committed:
Move XFEM code from framework to the newly-created xfem module ref #6320

bwspenc commented:
I moved the majority of the XFEM code into a module with this latest commit. I left the modifications that I made to FEProblem, Assembly, and Transient relatively unchanged. I'd like to focus initially on getting the interfaces to XFEM in those places to a state that everyone's happy with.

One thing to note here is that we're interested in generalizing what I'm doing in Transient to work with other models of fracture as well. We're solving the problem, checking for fracture growth, modifying something in the model if fracture growth should occur, and then repeating the solution until no new fracture growth should occur. Can we use the picard iteration mechanisms already there for multi-apps for this purpose, or is this a special case that needs its own separate mechanism?

permcody commented:
Thanks @bwspenc!

OK @idaholab/moose-developers, Let's find some time to look through this and get Ben some feedback.

Ben - Are there any new module dependencies that we need to work out (e.g. tensor mechanics -> xfem)? Thanks for this contribution, this is an exciting capability to add to the framework!

bwspenc commented:
@permcody I did make solid_mechanics depend on xfem because this adds a class in solid_mechanics that derives from a base class in xfem. That seems to work fine, but I was thinking afterwards that we could actually reverse that dependency and move that class into xfem if that helps anything.

friedmud commented:
@bwspenc if it's possible to reverse it... I would go that way.  That way everyone that compiles against solid mechanics doesn't need to compile XFEM as well.  Only those people who want XFEM will compile against it.

bwspenc commented:
@friedmud I just amended the commit to reverse the dependency, which works fine.  I agree that I prefer it that way (with xfem depending on solid_mechanics).
I also removed some example problems we had in there just to keep this cleaner at this point, and added a simple diffusion w/xfem test to the xfem module so that we're at least testing something in that module's tests. Most of the tests of xfem are being done in the combined module. By the way, I have a pretty complete set of unit tests of the element fragment algorithm that are in the moose/unit tests. I can move those to the xfem module, but it looks like the unit tests for all of the modules are currently being done in moose/unit so I followed that pattern.

permcody commented:
@friedmud - If you have time to look over this some more that would be awesome. This is a huge chunk of code to go through so there will be some bugs as we integrate it but we'll deal with that when it comes. I'm excited to have this capability in the framework!

permcody commented on framework/include/base/FEProblem.h:
```diff
@@ -1101,6 +1115,12 @@ class FEProblem :
+  XFEMInterface * _xfem;
+
+  ///Weight factors used by XFEM to modify qp weights for partial elements
+  std::map<dof_id_type, MooseArray<Real> > _xfem_weights;
```
Should this be a `VariableValue` type instead of a MooseArray?

bwspenc replied:
Whoops!  It doesn't actually matter what it should be because I should have deleted it during a refactor.

permcody commented on framework/src/base/FEProblem.C:
```diff
@@ -3682,8 +3720,9 @@ FEProblem::meshChanged()
   unsigned int n_threads = libMesh::n_threads();
 
-  for (unsigned int i = 0; i < n_threads; ++i)
+  for (unsigned int i = 0; i < n_threads; ++i){
```
bracket at the end of the line

permcody commented on framework/src/executioners/Transient.C:
```diff
@@ -406,19 +417,38 @@ Transient::solveStep(Real input_dt)
+         _problem.updateMeshXFEM() &&
+         (_xfem_update_count < _max_xfem_update))
+    {
+      Moose::out << "XFEM modifying mesh, repeating step"<<std::endl;
```
Use _console here to honor buffering and turning output off.

permcody commented on framework/src/executioners/Transient.C:
```diff
@@ -406,19 +417,38 @@ Transient::solveStep(Real input_dt)
+      {
+        _xfem_repeat_step = false;
+        _xfem_update_count = 0;
+        Moose::out << "XFEM not modifying mesh, continuing"<<std::endl;
```
Use _console

permcody commented on framework/include/base/XFEMInterface.h:
```diff
@@ -0,0 +1,112 @@
+   */
+  void setMesh(MeshBase* mesh)
+  {
+    _mesh=mesh;
```
spacing

bwspenc replied:
There are probably a ton of spacing issues to sort out.  Now I know what the policy is, I'll go through and clean all those up.

permcody commented on framework/include/base/XFEMInterface.h:
```diff
@@ -0,0 +1,112 @@
+   */
+  void setSecondMesh(MeshBase* mesh2)
+  {
+    _mesh2=mesh2;
```
spacing

permcody commented on framework/src/executioners/Transient.C:
```diff
@@ -82,6 +82,7 @@ InputParameters validParams<Transient>()
   params.addParamNamesToGroup("picard_max_its picard_rel_tol picard_abs_tol", "Picard");
 
   params.addParam<bool>("verbose", false, "Print detailed diagnostics on timestep calculation");
+  params.addParam<unsigned int>("max_xfem_update", std::numeric_limits<unsigned int>::max(), "Maximum number of times to update XFEM crack topology in a step due to evolving cracks");
```
Is this a sensible default? I'm not sure how often you are overriding this value but "unsigned int max" for practical purposes is unlimited.

bwspenc replied:
I think so.  The default should be that there is no limit.

permcody commented:
The header you've put on the XFEM files needs to be updated to the Modules version now. You should be able to use the "update_headers" script to accomplish this task in an automated way with a few tweaks.

dschwen commented on framework/src/base/Assembly.C:
```diff
@@ -1547,3 +1552,22 @@ Assembly::clearCachedJacobianContributions()
+void
+Assembly::modifyWeightsDueToXFEM(const Elem *elem)
+{
+  mooseAssert(_xfem != NULL, "This function should be called if xfem is inactive");
```
should **not** be called

dschwen commented on framework/src/base/FEProblem.C:
```diff
@@ -240,6 +240,8 @@ FEProblem::~FEProblem()
     delete it->second;
 
+  if (_xfem != NULL)
+    delete _xfem;
```
How about we use a smart pointer here. FEProblem deletes an object it did not allocate itself. This is just asking for trouble.

dschwen commented on framework/src/base/FEProblem.C:
```diff
@@ -3663,6 +3666,41 @@ FEProblem::adaptMesh()
+FEProblem::updateMeshXFEM()
+{
+  bool updated = false;
+  if (haveXFEM())
```
`_xfem != NULL` should be replaced by this call elsewhere as well.

bwspenc replied:
Done

dschwen commented on modules/xfem/README.md:
```diff
@@ -0,0 +1,6 @@
+Xfem
+=====
+
+"Fork Xfem" to create a new MOOSE-based application.
```
?! Stork leftovers?

dschwen commented on modules/xfem/include/actions/XFEMAction.h:
```diff
@@ -0,0 +1,45 @@
+/****************************************************************/
```
I guess these headers should be replaced by the MOOSE module header now.

bwspenc replied:
Done

dschwen commented on modules/xfem/include/actions/XFEMAction.h:
```diff
@@ -0,0 +1,45 @@
+  Real _xfem_crack_growth_increment;
+};
+
+#endif
```
`#endif //XFEMACTION_H`

dschwen commented on modules/xfem/include/actions/XFEMAction.h:
```diff
@@ -0,0 +1,45 @@
+  XFEMAction(InputParameters params);
+
+  virtual void act();
+protected:
```
add empty line above

dschwen commented on modules/xfem/include/auxkernels/XFEMCutPlaneAux.h:
```diff
@@ -0,0 +1,48 @@
+class XFEMCutPlaneAux : public AuxKernel
+{
+public:
+
```
remove empty line

dschwen commented on modules/xfem/include/auxkernels/XFEMCutPlaneAux.h:
```diff
@@ -0,0 +1,48 @@
+
+  /**
+   * Factory constructor, takes parameters so that all derived classes can be built using the same
+   * constructor.
```
Remove boilerplate comment

dschwen commented on modules/xfem/include/auxkernels/XFEMMarkerAux.h:
```diff
@@ -0,0 +1,47 @@
+  virtual Real computeValue();
+
+private:
+  XFEM *_xfem;
```
Does this need a writable pointer? If not I suggest making this `const`

bwspenc replied:
We can't set it in the initialization list, so yes, it does need to be writeable.

dschwen replied:
No.
http://stackoverflow.com/questions/1143262/what-is-the-difference-between-const-int-const-int-const-and-int-const
The member may have to be writable, but the object it points to probably
not.

On Mon, Feb 22, 2016 at 4:10 PM Ben Spencer <notifications@github.com>
wrote:

> In modules/xfem/include/auxkernels/XFEMMarkerAux.h
> <https://github.com/idaholab/moose/pull/6321#discussion_r53710067>:
>
> > +{
> > +public:
> > +
> > +  /**
> > +   * Factory constructor, takes parameters so that all derived classes can be built using the same
> > +   * constructor.
> > +   */
> > +  XFEMMarkerAux(const InputParameters & parameters);
> > +
> > +  virtual ~XFEMMarkerAux() {}
> > +
> > +protected:
> > +  virtual Real computeValue();
> > +
> > +private:
> > +  XFEM *_xfem;
>
> We can't set it in the initialization list, so yes, it does need to be
> writeable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6321/files#r53710067>.
>

dschwen commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -0,0 +1,194 @@
+#include "libmesh/vector_value.h"
+#include "libmesh/quadrature.h"
+
+enum XFEM_CUTPLANE_QUANTITY
```
I would not put these enums in the global namespace

dschwen commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -0,0 +1,194 @@
+public:
+
+  /**
+   * Constructor
```
Comment does not add value

dschwen commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -0,0 +1,194 @@
+  XFEM(MooseApp & app);
+
+  /**
+   * Destructor
```
neither does this ;-)

dschwen commented on modules/xfem/include/base/XFEMFuncs.h:
```diff
@@ -0,0 +1,59 @@
+
+bool line_exp_is_degenerate_nd(int dim_num, double p1[], double p2[]);
+
+int plane_normal_line_exp_int_3d(double pp[3], double normal[3], double p1[3], double p2[3], double pint[3]);
```
All these functions are in the global namespace and do not follow our naming conventions. Also array of double? You are including libmesh/point.h why not use that?

bwspenc replied:
These are functions that are taken from someone else's (LGPL'd) code, so I wanted to leave it somewhat intact. Yes, there's a lot that we should change about it. Maybe it's best to just start hacking it and acknowledge what we based it on.

dschwen replied:
Sure, if it is LGPL you are free to modify it, and doing so would clean the
code up quite abot and avoid the back and forth conversions

On Mon, Feb 22, 2016 at 4:12 PM Ben Spencer <notifications@github.com>
wrote:

> In modules/xfem/include/base/XFEMFuncs.h
> <https://github.com/idaholab/moose/pull/6321#discussion_r53710369>:
>
> > +void wissmannPoints(unsigned int nqp, std::vector<std::vector<Real> > &wss);
> > +
> > +void shapeFunc2D(unsigned int nen, std::vector<Real> &ss, std::vector<Point> &xl,
> > +                 std::vector<std::vector<Real> > &shp, Real &xsj, bool natl_flg);
> > +
> > +double r8vec_norm(int n, double a[]);
> > +
> > +void r8vec_copy(int n, double a1[], double a2[]);
> > +
> > +bool r8vec_eq (int n, double a1[], double a2[]);
> > +
> > +double r8vec_dot_product(int n, double a1[], double a2[]);
> > +
> > +bool line_exp_is_degenerate_nd(int dim_num, double p1[], double p2[]);
> > +
> > +int plane_normal_line_exp_int_3d(double pp[3], double normal[3], double p1[3], double p2[3], double pint[3]);
>
> These are functions that are taken from someone else's (LGPL'd) code, so I
> wanted to leave it somewhat intact. Yes, there's a lot that we should
> change about it. Maybe it's best to just start hacking it and acknowledge
> what we based it on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6321/files#r53710369>.
>

dschwen commented on modules/xfem/include/base/XFEM.h:
```diff
@@ -0,0 +1,194 @@
+
+  void addStateMarkedElem(unsigned int elem_id, RealVectorValue normal);
+  void addStateMarkedElem(unsigned int elem_id, RealVectorValue normal, unsigned int marked_side);
+  void addStateMarkedFrag(unsigned int elem_id, RealVectorValue normal);
```
I think those can and should be `const RealVectorValue &`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+  bool has_intersection = false;
+  double plane_point[3] = {_center(0), _center(1), _center(2)};
+  double plane_normal[3] = {_normal(0), _normal(1), _normal(2)};
+  double edge_point1[3] = {p1(0), p1(1), p1(2)};
```
Yeah, here you are explicitly converting ` Point` to a double array. I'd rather see the `plane_normal_line_exp_int_3d` function fixed and accept a `const Point &`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+  double cut_point[3] = {0.0,0.0,0.0};
+
+  if (plane_normal_line_exp_int_3d(plane_point, plane_normal, edge_point1, edge_point2, cut_point) == 1) {
+    Point temp_p(cut_point[0], cut_point[1], cut_point[2]);
```
And converting a double array back to a Point...

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+{
+  Point p1_to_p2 = p2 - p1;
+  Point p_to_p1 = p1 - p;
+  Point p_to_p2 = p2 - p;
```
these temporary expression names are longer than the terms they represent. Just ditch them.

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+  Point p_to_p2 = p2 - p;
+  Real dotp1 = p_to_p1*p1_to_p2;
+  Real dotp2 = p_to_p2*p1_to_p2;
+  if (dotp1*dotp2 <= 0.0)
```
toss this if/else construct and just `return dotp1 * dotp2 <= 0.0;`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+XFEMGeometricCut3D::getRelativePosition(Point p1, Point p2, Point p)
+{
+  // get the relative position of p from p1
+  Real full_len = std::sqrt((p2 - p1).size_sq());
```
No. Just use `size()` rather than taking the square root youself!

dschwen commented on modules/xfem/src/base/XFEMSquareCut.C:
```diff
@@ -0,0 +1,67 @@
+#include "XFEMFuncs.h"
+#include "EFAFuncs.h"
+
+XFEMSquareCut::XFEMSquareCut(std::vector<Real> square_nodes):
```
This could just take a vector of Points. Then you do not have to convert stuff around, plus the code would be more expressive.

dschwen commented on modules/xfem/src/base/XFEMSquareCut.C:
```diff
@@ -0,0 +1,67 @@
+
+  for (unsigned int i = 0; i < 4; ++i)
+  {
+    unsigned int iplus1(i < 3 ? i+1 : 0);
```
to initialize primitive types we use the `=` operator

bwspenc replied:
Sure, but then we couldn't use the ternary operator here, so this would take multiple lines of code. We do this all over the place. This seems better to me, no?

dschwen replied:
Huh?! That is not correct. You simply write
```
const unsigned int iplus1 = i < 3 ? i + 1 : 0;
```

Even better would be to ditch this line and write `Point ray2 = _vertices[(i + 1) % 4] - _center;` below.

dschwen commented on modules/xfem/src/base/XFEMSquareCut.C:
```diff
@@ -0,0 +1,67 @@
+  {
+    unsigned int iplus1(i < 3 ? i+1 : 0);
+    Point ray1 = _vertices[i] - _center;
+    Point ray2 = _vertices[iplus1] - _center;
```
these type of shorthands should be `const ` IMO

dschwen commented on modules/xfem/src/base/XFEMGeometricCut.C:
```diff
@@ -0,0 +1,45 @@
+  {
+    if (time >= t_end)
+    {
+      fraction = 1.0;
```
We omit curly braces around single statements.

dschwen commented on modules/xfem/src/base/XFEMGeometricCut2D.C:
```diff
@@ -0,0 +1,167 @@
+
+XFEMGeometricCut2D::XFEMGeometricCut2D(Real x0_, Real y0_, Real x1_, Real y1_, Real t_start_, Real t_end_):
+  XFEMGeometricCut(t_start_, t_end_),
+  x0(x0_),
```
The trailing `_` should go, instead we use a leading `_` for member variables.

dschwen commented on modules/xfem/src/base/XFEMGeometricCut2D.C:
```diff
@@ -0,0 +1,167 @@
+  for (unsigned int i = 0; i < nsides; ++i)
+  {
+    Node *node1 = elem->get_node(i);
+    Node *node2 = elem->get_node(i<(nsides-1) ? i+1 : 0);
```
should be const IMO

dschwen commented on modules/xfem/src/base/XFEMGeometricCut2D.C:
```diff
@@ -0,0 +1,167 @@
+
+    Real cut_dir_cross_seg_dir = crossProduct2D(cut_dirx, cut_diry, seg_dirx, seg_diry);
+
+    if (fabs(cut_dir_cross_seg_dir) > tol)
```
please use `std::abs` instead of `fabs`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut2D.C:
```diff
@@ -0,0 +1,167 @@
+                            cut_dir_cross_seg_dir;
+        if (seg_int_frac >= 0.0 && seg_int_frac <= 1.0) //TODO: revisit end cases for intersections with corners
+        {
+
```
empty line

dschwen commented on modules/xfem/src/base/XFEMGeometricCut2D.C:
```diff
@@ -0,0 +1,167 @@
+  bool cut_frag = false;
+  Real tol = 1.e-15;
+
+  Real fraction = cutFraction(time);
```
`const`... everything `const`!!!

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+
+XFEMGeometricCut3D::XFEMGeometricCut3D(Real t0, Real t1):
+  XFEMGeometricCut(t0,t1),
+  _center(Point(0.0,0.0,0.0)),
```
The default constructor for `Point` initializes the member as null. so this should be `_center(),`!

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+  bool cut_elem = false;
+
+   int hex_indices[6][4] = {{0,3,2,1},{0,1,5,4},{1,2,6,5},{2,3,7,6},{3,0,4,7},{4,5,6,7}};
+   int tet_indices[4][3] = {{0,2,1},{0,1,3},{1,2,3},{2,0,3}};
```
make `const` (it helps the reader and the compiler)

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+   else if (elem->n_nodes() == 4)
+     n_face_nodes = 3;
+   else
+     mooseError("this method only works for lienar hexes and tets");
```
You are _not_ checking the element type here, just the number of nodes. This could be dangerous.

bwspenc replied:
I agree completely.  There's a method on the fe class that will tell us that, but not on the elem.  Any idea on how to get the fe associated with an elem when that's all we have in our hand?

permcody replied:
You should be able to do this right from the Elem (n_face_nodes), I don't believe you need the FE object. Tagging in @jwpeterson in case you need further clarification.

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+   else
+     mooseError("this method only works for lienar hexes and tets");
+
+    for (unsigned int i=0; i < elem->n_sides(); i++)
```
operator spacing (and `++i`)

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+        {
+          node_id1 = hex_indices[i][j];
+          node_id2 = hex_indices[i][jplus1];
+        }else if (elem->n_nodes() == 4) // tet
```
line break after `}`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+        Point p1((*node1)(0), (*node1)(1), (*node1)(2));
+        Point p2((*node2)(0), (*node2)(1), (*node2)(2));
+
+        Point pint(0.0,0.0,0.0);
```
again: no need to explicitly call the CTOR with 0,0,0, just use the default constructor!

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+        if (intersectWithEdge(p1,p2,pint))
+        {
+          cut_edges.push_back(j);
+          cut_pos.push_back(getRelativePosition(p1,p2,pint));
```
We insert a space after the `,` operator

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+      if (cut_edges.size() == 2)
+      {
+       cut_elem = true;
+       cutFace mycut;
```
That type should be named `CutFace`

dschwen commented on modules/xfem/src/base/XFEMGeometricCut3D.C:
```diff
@@ -0,0 +1,168 @@
+}
+
+bool
+XFEMGeometricCut3D::cutFragmentByGeometry(std::vector<std::vector<Point> > & /*frag_edges*/,
```
Do both arguments need to be writable references?

bwspenc replied:
Yes

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+
+EFAEdge::EFAEdge(EFANode * node1, EFANode * node2):
+  _edge_node1(node1),
+  _edge_node2(node2)
```
initializer lists are indented by 4 spaces (and a space in front of the `:`)

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+}
+
+EFAEdge::~EFAEdge() // do not delete edge node - they will be deleted
+{}                  // in EFAelement's destructor
```
Yeah, I'm not sure I like the sound of that ;-). Smart pointers are most likely the better way here.

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+  bool partially_overlap = false;
+  if (containsEdge(other) || other.containsEdge(*this))
+    partially_overlap = true;
+  return partially_overlap;
```
this whole function body could be written as
```
return containsEdge(other) || other.containsEdge(*this);
```

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+  if (containsNode(other._edge_node1) &&
+      containsNode(other._edge_node2))
+    contains = true;
+  return contains;
```
same pattern here

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+    master_weights.push_back(1.0);
+    masters_found = true;
+  }
+  else if (_edge_node2 == node)
```
the cases can be combined
```
  if (_edge_node1 == node || _edge_node2 == node)
  {
    master_nodes.push_back(node);
    master_weights.push_back(1.0);
    masters_found = true;
  }
```

dschwen replied:
You might want to check if the first weight should really be `1.0` and not `0.0`.

bwspenc replied:
It should.

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+  if (index == 0)
+    return _edge_node1;
+  else if (index == 1)
+    return _edge_node2;
```
might as well make _edge_node a two element array then...

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+{
+  EFANode* tmp = _edge_node1;
+  _edge_node1 = _edge_node2;
+  _edge_node2 = tmp;
```
`std::swap`

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+bool
+EFAEdge::hasIntersection() const
+{
+  return (_embedded_nodes.size() > 0);
```
parens are not necessary

bwspenc replied:
The Code Standards don't say anything about that.  Is it a problem to have them?

dschwen replied:
Well yeah, it is a problem, because it makes the expression unnecessarily complicated. If you can write it simpler you should. Having too many ways to express the same thing makes the code look messy and harder to read. So the obvious choice should always be the simplest form.

bwspenc replied:
We need to update the Code Standards to prohibit this then.

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+unsigned int
+EFAEdge::getEmbeddedNodeIndex(EFANode * node) const
+{
+  unsigned int index = 99999;
```
no need to set this to 99999

bwspenc replied:
Code checkers could complain about it being an uninitialized variable.

dschwen replied:
No, they would be wrong (and I bet they are smart enough anyways). Code checkers analyze the possible code paths and only complain if a variable is _used_ while uninitialized. This is not the case here.

permcody replied:
Yes, please change this. I went through all of MOOSE and took out all the other random number of 9 initializations. We don't need this as Daniel mentioned. You can verify this with "make sa" yourself.

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+{
+  unsigned int index = 99999;
+  bool have_index = false;
+  double tol = 1.e-4;
```
should be `const`

dschwen commented on modules/xfem/src/efa/EFAEdge.C:
```diff
@@ -0,0 +1,419 @@
+      isEmbeddedNode(node))
+    return true;
+  else
+    return false;
```
`return _edge_node1 == node || _edge_node2 == node || isEmbeddedNode(node);`

dschwen commented on modules/xfem/src/efa/EFAElement.C:
```diff
@@ -0,0 +1,366 @@
+        else
+        {
+          EFAError("Attempting to merge nodes: "<<childNode->id()<<" and "
+                      <<childOfNeighborNode->id()<<" but both are permanent themselves");
```
`<<` is a binary operator which we pad with whitespace

dschwen commented on modules/xfem/include/efa/EFAFuncs.h:
```diff
@@ -0,0 +1,83 @@
+  {
+    delete i->second;
+    theMap.erase(i);
+    elemToDelete=NULL;
```
this does nothing

dschwen commented on modules/xfem/include/efa/EFAFuncs.h:
```diff
@@ -0,0 +1,83 @@
+  if (i != theMap.end())
+  {
+    delete i->second;
+    theMap.erase(i);
```
erade destroys the removed element, so you used smart pointers the delete stuff would be unnecesary

bwspenc replied:
I'm hesitant to use smart pointers for this code because this algorithm has no dependencies on anything other than the standard library, and I'd like to keep it that way unless there's a really compelling reason.

dschwen replied:
Are you ever going to pull this code back out? If so you could add your own shared pointer wrapper. You could also simply require C++11 (we are going that route sooner or later anyways)!
The compelling reason here is that for such a complex piece of code the probability you are going to leak memory sooner or later is pretty big.

bwspenc replied:
As the code is structured, it would be trivial to pull this algorithm out and use it in another code, so I don't want to rule that possibility out. We need to talk offline about what the options for smart pointers would be.

dschwen commented on modules/xfem/src/efa/EFAElement.C:
```diff
@@ -0,0 +1,366 @@
+      if (!deleteFromMap(TempNodes, childNode))
+      {
+        EFAError("Attempted to delete node: "<<childNode->id()<<" from TempNodes, but couldn't find it");
+      }
```
no curly braces around single statement bodies

dschwen commented on modules/xfem/src/efa/EFAElement.C:
```diff
@@ -0,0 +1,366 @@
+      childNode = newNode;
+    }
+  }
+}
```
and here ends the diff on Github. I suggest addressing the comments we gave so far and splitting this into multiple (two) PRs

bwspenc replied:
Hmm.  I'll have to ponder how I could split it into two PRs and still have it work.  Maybe I can do that.

dschwen commented on modules/xfem/src/efa/EFAElement.C:
```diff
@@ -0,0 +1,366 @@
+  for (unsigned int i = 0; i < _crack_tip_neighbors.size(); ++i)
+  {
+    if (_crack_tip_neighbors[i] == neighbor_index)
+      crack_tip_neighbor_exist = true;
```
you can add a `break;` here

permcody commented on modules/xfem/src/base/XFEMFuncs.C:
```diff
@@ -0,0 +1,597 @@
+    mooseError("ShapeFunc2D only works for linear quads and tris!");
+}
+
+double r8vec_norm(int n, double a[])
```
I agree with @dschwen on this. At the very least all of these functions should be put in a Namespace especially if you are hesitant to match our coding conventions. We do allow deviations from conventions for "contrib" codes. If you don't want to change it, how about putting them in contrib?

As far as wrappers go, it would be easy to have an XFem header that defines the wrapper now (just a few typedefs will do until you pull it out). You could define your own "Real" type too so it matches MOOSE. That may not be necessary if you put all of this in contrib though. I just don't like seeing it mixed in with the module code.

bwspenc replied:
I put these functions in a namespace.

permcody commented:
This is good and we are close to merging. I'd like to see several tickets opened for the the comments that you aren't going to address right now. If we have tickets open, we'll at least capture that there is work left to do to clean this up.

dschwen commented on modules/xfem/src/base/XFEMCircleCut.C:
```diff
@@ -0,0 +1,48 @@
+}
+
+XFEMCircleCut::~XFEMCircleCut()
+{}
```
Style guide states these braces should be on separate lines.

bwspenc commented:
I put in the following tickets for known issues to resolve: #6417, #6418, #6419, #6420, #6421, #6422. These are all issues that I believe can be resolved later after this is merged. Otherwise, I think this PR is ready to merge.

permcody commented:
Thanks Ben!

jiangwen84 commented:
@bwspenc  issue #6418 is not related to XFEM. :-)

bwspenc commented:
jiangwen84 -- good catch -- I looked a the list and I thought I had a consecutive set, but one got slipped in there when I was typing these.  I'm editing the comment to remove it.

permcody commented:
Alright, I'm merging this. We'll turn on coverage in the new module (@brianmoose) and start working on all the tickets you opened up. Thanks!

bwspenc commented:
Awesome!  Thanks for all of the help with this. On to the next steps!

permcody commented:
Ben,

Would you like to put out an announcement to the list about this new
module? Also, I wouldn't mind if you created a blog post on
mooseframework.org. I can put the text up there if you write something up.

Thanks

On Thu, Feb 25, 2016 at 1:57 PM, Ben Spencer <notifications@github.com>
wrote:

> Awesome! Thanks for all of the help with this. On to the next steps!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/6321#issuecomment-188983190>.
>

jiangwen84 commented:
@bwspenc If you decide to put an announcement, I can help you add a XFEM 'funny' picture.

bwspenc commented:
@permcody Yes, I'll put out an announcement in the next few days.

brianmoose commented:
Coverage is at https://mooseframework.org/docs/coverage/xfem/xfem_coverage/
Currently at 73.7%
